frame/support/src/traits/hooks.rs

```diff
@@ -239,64 +190,111 @@ impl OnRuntimeUpgrade for Tuple {
+	/// See [`Hooks::integrity_test`].
 	fn integrity_test() {}
 }
 
-/// The pallet hooks trait. Implementing this lets you express some logic to execute.
+/// The pallet hooks trait. This is merely an umbrella trait for:
+///
+/// - [`OnInitialize`]
+/// - [`OnFinalize`]
+/// - [`OnRuntimeUpgrade`]
+/// - [`OffchainWorker`]
+/// - [`OnIdle`]
+/// - [`IntegrityTest`]
+///
+/// ## Ordering
+///
+/// For all hooks, except [`OnIdle`] the order of execution is derived from how the pallets are
+/// ordered in [`crate::construct_runtime`].
+///
+/// ## Summary
+///
+/// In short, the following diagram shows the
 pub trait Hooks<BlockNumber> {
-	/// The block is being finalized. Implement to have something happen.
+	/// Block initialization hook. This is called at the very beginning of block execution.
+	///
+	/// Must return the non-negotiable weight of both itself, and whatever [`on_finalize`] wishes to
+	/// consumed.
+	///
+	/// The weight returned by this is treated as [`DispatchClass::Mandatory`], meaning that it MUST
+	/// BE EXECUTED. This implies that this hook should only ever be used if the logic in it must
+	/// happen under any circumstance. For all similar use cases where a delay is acceptable, use
+	/// [`poll`].
+	///
+	/// NOTE: This function is called BEFORE ANY extrinsic in a block is applied, including inherent
+	/// extrinsics. Hence for instance, if you runtime includes `pallet-timestamp`, the `timestamp`
+	/// is not yet up to date at this point.
+	fn on_initialize(_n: BlockNumber) -> Weight {
+		Weight::zero()
+	}
+
+	/// Similar to [`on_initialize`], but:
+	///
+	/// 1. is called after [`on_initialize`] of all pallets is done executing,
+	/// 2. execution on *each block* is not guaranteed
+	///
+	/// The latter point is usually correlated with the block resources being available, for example
+	/// `Weight` leftover.
+	fn poll(_n: BlockNumber) -> Weight {
+		Weight::zero()
+	}
+
+	/// Block finalization hook. This is called at the very end of block execution.
+	///
+	/// Note that this has nothing to do with finality in the "consensus" sense.
+	///
+	/// Note that the non-negotiable weight for this has must have already been returned by
+	/// [`OnInitialize`]. It usage along is not permitted.
+	///
+	/// Similar to [`OnInitialize`] it should only be used when execution is absolutely necessary.
 	fn on_finalize(_n: BlockNumber) {}
 
-	/// This will be run when the block is being finalized (before `on_finalize`).
+	/// Hook to consume a block's idle time. This will run when the block is being finalized (before
+	/// [`on_finalize`]).
 	///
-	/// Implement to have something happen using the remaining weight. Will not fire if the
-	/// remaining weight is 0.
+	/// Given that all dispatchables are already executed and noted, and the weight for
+	/// [`on_finalize`], which comes next, is also already accounted for (via `on_initialize`), this
+	/// hook consumes anything that is leftover.
```

Comment:
I get confused with many commas, maybe something like this?

```suggestion
	/// Given that all dispatchables are already executed and noted (and the weight for
	/// [`on_finalize`], which comes next, is also already accounted for via `on_initialize`), this
	/// hook consumes anything that is leftover.
```